design/20230302.core-issuers.md

```diff
@@ -0,0 +1,70 @@
+several large and complex public cloud APIs (GCP, AWS, Azure, DigitalOcean, Cloudflare and lots more) and some external
+SDKs (including Hashicorp Vault and Venafi VCert) due to it containing the "built-in" issuers alongside the ACME DNS-01
+solvers.
+
+While the `controller` component acts as all built-in issuers, that's not its only job. It also handles approval,
+triggering reissuance, key management and various other tasks through its [long list](https://github.com/cert-manager/cert-manager/blob/4e889b702b8bbfb082b7a3234569dc173c1c286d/cmd/controller/app/options/options.go#L186-L206)
+of Kubernetes controllers enabled by default.
+
+This large list of tasks has several side effects:
+
+- The `controller` component has many complex dependencies related to issuance which can be hard to upgrade
+- The many tasks it carries out introduce an increased attack surface, where one controller could be tricked into doing
+  something malicious to another part of the controller.
+- RBAC for the `controller` is very complex with wide-ranging permissions, again increasing risks if there are any
+  compromises or vulnerabilities.
+
+In addition, having so much logic in the `controller` makes it difficult to disable functionality for security or
+efficiency purposes. If a users know they'll only ever use one particular external issuer, it's reasonable for them to
+want to disable the built-in issuers. The current controller makes doing so hard.
+
+## Proposed Solution: `core-issuers` Component
+
+**In short:** Remove all issuer-related functionality from `controller` and add them to a new `core-issuers` component
+
+As a solution, we could separate the controllers relating to issuance from the core `controller` and add a new
+component called `core-issuers` which would run as a separate pod. It would reconcile over `Issuer` and `ClusterIssuer`
+resources and would sign `CertificateRequest` and `CertificateSigningRequest` resources which use the built-in issuers.
+
+This new component would:
+
+- Be simple to disable
+- Isolate issuer-related dependencies from core functionality like triggering renewal
+- Reduce the complexity of the `controller` significantly
+- Allow RBAC for the `controller` to be reduced
+
+This component would run as a new `Deployment`, alongside the existing cert-manager pods.
+
+Much of the code for the new component would be copied almost verbatim from the existing `controller` - they'd differ
+in which Kubernetes controllers they start, but both would essentially be binaries which start a list of controllers.
+
+## Potential Problems
+
+### Metrics / Logs
+
+Users who have dashboards consuming metrics from the `controller` component will currently see metrics for all issuers,
+and if we moved to a `core-issuers` component those metrics would "disappear" until the user reconfigures their
+dashboards. Similarly, logs relating to issuance would "move".
```

Comment:
If the metrics are exposed with the same names then I think a lot of dashboards would still show them. Most setups collect metrics from all pods, and then the dashboards will show particular metric names and labels. Some of the labels will change with this move, but I imagine that most dashboards won't be filtering on those labels.

Reply:
> Most setups collect metrics from all pods, and then the dashboards will show particular metric names and labels. 

Ah that's really useful context, thanks! That removes a huge potential pain point.